### PR TITLE
feat: add `storage` module

### PR DESCRIPTION
**Pull request by lemeurherve**
This PR adds a module to create a service principal to manipulate Azure File Shares.

Ref:
- https://github.com/jenkins-infra/helpdesk/issues/3414#issuecomment-1856196680

### CODE REVIEW

File: terraform/modules/azure-jenkinsinfra-storage/variables.tf [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+# Optional variables
+variable "default_tags" {
+  type    = map(string)
+  default = {}
+}
+
+variable "storage_active_directory_url" {
+  type    = string
+  default = "https://github.com/jenkins-infra/azure"
```
Comment (dduportal) [inline terraform/modules/azure-jenkinsinfra-storage/variables.tf:27]:
```suggestion
  default = ""
```

or 

```suggestion
  default = null
```

as we can call this from different repositories

**Review by dduportal — COMMENTED** [review]
Do you have a (draft) PR with code showing how do you "call" this module (the PR checks are expected to fail of course) to help in the review (while reviewing as "consumer")

**Comment by lemeurherve** [comment]
> Do you have a (draft) PR with code showing how do you "call" this module (the PR checks are expected to fail of course) to help in the review (while reviewing as "consumer")

Opened https://github.com/jenkins-infra/azure/pull/557

File: terraform/modules/azure-jenkinsinfra-storage/variables.tf [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Required variables
+variable "service_fqdn" {
+  type = string
+}
+
+variable "storage_service_principal_ids" {
```
Comment (dduportal) [inline terraform/modules/azure-jenkinsinfra-storage/variables.tf:6]: See my comment in https://github.com/jenkins-infra/azure/pull/557/files#r1431730896 which underline the poor naming I did on the controller module.

WDYT about the following name in this case to make it easier to understand?

```suggestion
variable "storage_active_directory_application_owners" {
```

File: terraform/modules/azure-jenkinsinfra-storage/variables.tf [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+variable "storage_service_principal_end_date" {
+  type = string
+}
+
+variable "storage_file_share_ids" {
+  type    = list(string)
+  default = []
+}
```
Comment (dduportal) [inline terraform/modules/azure-jenkinsinfra-storage/variables.tf:17]:
```suggestion
variable "storage_file_share_id" {
  type    = string
}
```

- Singular form (not sure to see the case for multiple filestorages: I would use different SPs instead) but not sure about this
- Whatever is the outcome of the above point, this variable should be mandatory (e.g. no default)

File: terraform/modules/azure-jenkinsinfra-storage/variables.tf [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+# Optional variables
+variable "default_tags" {
+  type    = map(string)
+  default = {}
+}
+
+variable "storage_active_directory_url" {
+  type    = string
+  default = ""
```
Comment (dduportal) [inline terraform/modules/azure-jenkinsinfra-storage/variables.tf:27]:
```suggestion
```

WDYt about making this variable mandatory?

**Review by dduportal — CHANGES_REQUESTED** [review]
This is really a nice job!

I've provided a few suggestion, with one blocker.

Additionally, how do you feel about renaming the module to underline its intent?

I would expect `azure-jenkinsinfra-storage` to create the file storage for me while it only sets up an Azure SP and its permissions to write to this filestorage.

WDYT about `azure-jenkinsinfra-storage-serviceprincipal-writer`?

File: terraform/modules/azure-jenkinsinfra-storage/variables.tf (hunk 3 above)
Comment (lemeurherve) [reply terraform/modules/azure-jenkinsinfra-storage/variables.tf:17]: I've reproduced the plural form used for packer RG ids in https://github.com/jenkins-infra/shared-tools/blob/23ec7072579d46feca3893b262c5bef7bd91acd8/terraform/modules/azure-jenkinsinfra-controller/variables.tf#L100-L103

Singular form is fine to me.

**Review by smerle33 — APPROVED** [review]
LGTM 👍

**Pull request merged**